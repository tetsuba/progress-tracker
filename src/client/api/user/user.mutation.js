import gql from 'graphql-tag'

export const VERIFY_EMAIL_MUTATION = gql`
  mutation($input: UserEmailInput!) {
    verifyEmail(input: $input) {
      confirmation
    }
  }
`

export const SEND_PASSWORD_RESET_CONFIRMATION_MUTATION = gql`
  mutation($input: UserEmailInput!) {
    sendPasswordResetConfirmation(input: $input) {
      confirmation
    }
  }
`

export const LOGIN_MUTATION = gql`
  mutation($input: LoginUserInput!) {
    userLogin(input: $input) {
      id
      firstName
      lastName
      token
      email
    }
  }
`

export const REGISTER_USER_MUTATION = gql`
  mutation NewUser($input: NewUserInput!) {
    newUser(input: $input) {
      id
    }
  }
`

export const REST_PASSWORD_MUTATION = gql`
  mutation ResetPassword($input: ResetPasswordInput!) {
    resetPassword(input: $input) {
      confirmation
    }
  }
`
