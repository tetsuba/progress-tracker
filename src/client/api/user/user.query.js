import gql from 'graphql-tag'

export const GET_USER_DETAILS_QUERY = gql`
  {
    getUserDetails {
      firstName
      lastName
      email
    }
  }
`

export const IS_USER_SESSION_EXPIRED = gql`
  {
    isUserSessionExpired {
      success
    }
  }
`

export const VALIDATE_USER_EMAIL_QUERY = gql`
  query ValidateUserEmail($token: String!) {
    validateUserEmail(token: $token) {
      success
    }
  }
`
