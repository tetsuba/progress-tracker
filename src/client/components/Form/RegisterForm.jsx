import React from 'react'
import { Button, Col, Form } from 'react-bootstrap'

// COMPONENTS
import Box from '../Box/Box'
import PasswordInputs from './PasswordInputs'

// HOOKS
import { useInputChange } from '../../hooks/hooks'

// UTILS
import { passwordsDoNotMatched } from './form-utils'

export default function RegisterForm(props) {
  const { handleSubmit, error } = props
  const errorMessage = error && error.graphQLErrors[0].message

  const [inputs, setInputs] = useInputChange({
    firstName: '',
    lastName: '',
    email: '',
    newPassword: '',
    confirmPassword: '',
  })

  return (
    <Box max={500}>
      <h3>Registration</h3>
      <Form
        id="RegisterForm"
        className="mt-4 mb-5"
        onSubmit={(e) => {
          e.preventDefault()
          const options = {
            variables: {
              input: {
                firstName: inputs.firstName,
                lastName: inputs.lastName,
                email: inputs.email,
                password: inputs.newPassword,
              },
            },
          }
          handleSubmit(options)
        }}
      >
        <Form.Row>
          <Form.Group as={Col} controlId="RegisterFirstName">
            <Form.Control
              placeholder="First Name"
              onChange={setInputs}
              value={inputs.firstName}
              name="firstName"
              required
            />
          </Form.Group>
        </Form.Row>
        <Form.Row>
          <Form.Group as={Col} controlId="RegisterLastName">
            <Form.Control
              placeholder="Last Name"
              onChange={setInputs}
              value={inputs.lastName}
              name="lastName"
              required
            />
          </Form.Group>
        </Form.Row>
        <Form.Row className="mb-3">
          <Form.Group as={Col} controlId="RegisterEmail">
            <Form.Control
              type="email"
              placeholder="Your email address"
              onChange={setInputs}
              value={inputs.email}
              name="email"
              required
              isInvalid={!!errorMessage}
            />
            <Form.Control.Feedback type="invalid">
              {errorMessage}
            </Form.Control.Feedback>
          </Form.Group>
        </Form.Row>

        <PasswordInputs inputs={inputs} setInputs={setInputs} register />

        <Button
          id="RegisterFormSubmit"
          variant="primary"
          type="submit"
          className="float-right"
          disabled={passwordsDoNotMatched(inputs)}
        >
          Register
        </Button>
      </Form>
    </Box>
  )
}
