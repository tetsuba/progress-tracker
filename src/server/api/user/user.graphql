
type UserType {
    id: String
    firstName: String
    lastName: String
    email: String
    password: String
    country: String
}

type UserLoginType {
    id: String
    firstName: String
    lastName: String
    token: String
    email: String
}

type UserVerifyType {
    confirmation: String
}

input NewUserInput {
    firstName: String
    lastName: String
    email: String
    password: String
    country: String
}

input LoginUserInput {
    email: String
    password: String
}

input UserEmailInput {
    email: String
}

input UserDataInput {
    id: String
    firstName: String
    lastName: String
}

input ResetPasswordInput {
    id: String
    password: String
}

extend type Query {
    getUserData: UserType!
    isUserSessionExpired(token: String): UserType!
}

type Mutation {
    newUser(input: NewUserInput!): UserType!
    userLogin(input: LoginUserInput!): UserLoginType!
    updateUserData(input: UserDataInput!): UserLoginType!
    verifyEmail(input: UserEmailInput!): UserVerifyType!
    sendPasswordResetConfirmation(input: UserEmailInput!): UserVerifyType!
    resetPassword(input: ResetPasswordInput!): UserVerifyType!
}
