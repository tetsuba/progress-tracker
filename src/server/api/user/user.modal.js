const mongoose = require('mongoose')

const userSchema = new mongoose.Schema(
  {
    firstName: {
      type: String,
      required: true,
    },
    lastName: {
      type: String,
      required: true,
    },
    email: {
      type: String,
      required: true,
    },
    password: {
      type: String,
      required: true,
    },
    country: {
      type: String,
      required: false,
    },
    isVerified: {
      type: Boolean,
      default: false,
    },
  },
  { _id: true }
)

module.exports = userSchema
