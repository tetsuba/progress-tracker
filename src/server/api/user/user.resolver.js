const { validateEmail } = require('../../utils/common')

const {
  createNewUser,
  findUser,
  updateUser,
  findEmail,
  resetPassword,
} = require('./user.CRUD')
const { getEmailMailOptions } = require('../../utils/email/sendEmail.js')
const { errorName } = require('../../errorHandling')
const { createVerificationToken } = require('../../utils/token')
const { passwordsDoNotMatch } = require('../utils')

module.exports = {
  Query: {
    isUserSessionExpired: async (_, args, context) => {
      try {
        return await findUser({
          email: context.user.email,
          password: context.user.password,
        })
      } catch (e) {
        return e
      }
    },
    getUserData: async (_, args, context) => {
      console.log('@@@@@@@@@@@@ getUserData: ', context)
      try {
        return await findUser({
          email: context.user.email,
          password: context.user.password,
        })
      } catch (e) {
        return e
      }
    },
  },

  Mutation: {
    newUser: async (_, args, context, info) => {
      const name = 'confirmEmail'

      if (validateEmail(args.input.email)) {
        throw new Error(errorName.NOT_VALID_EMAIL)
      }

      try {
        const mailOptions = getEmailMailOptions(name)
        return await createNewUser(args.input, mailOptions)
      } catch (err) {
        return err
      }
    },

    userLogin: async (_, args, context, info) => {
      const { User, Token } = context.models
      const {
        input: { email, password },
      } = args
      const user = await User.findOne({ email })

      /* Error Handling:
       * - User is not found
       * - Passwords do not match
       *  */
      if (!user || passwordsDoNotMatch(password, user.password)) {
        throw new Error(errorName.INCORRECT_USER_DETAILS)
      }

      /* Error Handling:
       * - User email has not been verified
       *  */
      if (!user.isVerified) {
        throw new Error(errorName.EMAIL_NOT_VERIFIED)
      }

      /* Create a session token (12hrs):
       *  */
      const verificationToken = createVerificationToken(user)
      const { token } = await Token.create({
        _userId: user._id,
        token: verificationToken,
      })

      return { token }
    },

    updateUserData: async (_, args, context, info) => {
      try {
        return await updateUser(args.input)
      } catch (err) {
        return err
      }
    },

    verifyEmail: async (_, args, context) => {
      const name = 'confirmEmail'
      try {
        const mailOptions = getEmailMailOptions(name)
        return await findEmail({ email: args.input.email }, mailOptions)
      } catch (err) {
        if (err.name === 'TypeError') {
          console.log(
            `Property "name": "${name}" is incorrect and can not get mail options`
          )
          return new Error(errorName.INTERNAL_SERVER_ERROR)
        }
        return err
      }
    },

    sendPasswordResetConfirmation: async (_, args, context) => {
      const name = 'resetPassword'
      try {
        const mailOptions = getEmailMailOptions(name)
        return await findEmail({ email: args.input.email }, mailOptions)
      } catch (err) {
        if (err.name === 'TypeError') {
          console.log(
            `Property "name": "${name}" is incorrect and can not get mail options`
          )
          return new Error(errorName.INTERNAL_SERVER_ERROR)
        }
        return err
      }
    },

    resetPassword: async (_, args, context) => {
      try {
        return await resetPassword(args.input.token, args.input.password)
      } catch (err) {
        return err
      }
    },
  },
}

// NOTES
// fieldName: (parent, args, context, info) => data;

// parent:  An object that contains the result returned from the resolver on the parent type
// args:    An object that contains the arguments passed to the field
// context: An object shared by all resolvers in a GraphQL operation. We use the context to contain per-request state such as authentication information and access our data sources.
// info:    Information about the execution state of the operation which should only be used in advanced cases
