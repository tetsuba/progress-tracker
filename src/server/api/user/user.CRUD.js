const { getUserId, deleteToken } = require('../token/token.CRUD');
const jwt = require('jsonwebtoken');

// This order can not be changed
const mongoose = require('mongoose');
require('./user.modal');
const User = mongoose.model('user');
// -----------------------------

const { createToken } = require('../token/token.CRUD');
const { sendMail } = require('../../utils/email/sendEmail');
const { errorName } = require('../../errorHandling');
const { createAuthToken } = require('../../utils/token');

async function findEmail(data, mailOptions) {
    try {
        const user = await User.findOne({ email: data.email });

        // ERROR - User not found
        if (!user) {
            throw new Error(errorName.EMAIL_DOES_NOT_EXIST)
        }

        // ERROR - User is verified
        if (user.isVerified && mailOptions.subject.includes('confirm')) {
            throw new Error(errorName.EMAIL_VERIFIED);
        }

        // Do not send an email when running functional tests
        if (!process.env.TEST) {
            const obj = await createToken(user);
            await sendMail(obj.token, mailOptions);
        }

        return {
            confirmation: 'Please check your email and confirm.'
        }

    } catch(error) {
        return error
    }

}

async function findUser(data) {
    try {
        const user = await User.findOne({ email: data.email });

        // ERROR - User is not found or password does not match
        if (!user || user.password !== data.password) {
            throw new Error(errorName.INCORRECT_USER_DETAILS)
        }

        // ERROR - Email address not verified
        if (!user.isVerified) {
            throw new Error(errorName.EMAIL_NOT_VERIFIED)
        }

        const token = createAuthToken(user);

        const payload = jwt.verify(
            token,
            process.env.REACT_APP_JWT_AUTH_SECRET
    )

        return {
            firstName: user.firstName,
            lastName: user.lastName,
            email: user.email,
            id: user._id,
            token
        }
    } catch(error) {
        return error
    }
}

// RegisterUser
async function createNewUser(data) {
    try {
        const user = await User.findOne({ email: data.email});

        // ERROR - User is not found or password does not match
        if (user) {
            throw new Error(errorName.EMAIL_ALREADY_EXIST)
        }

        const newUser = await User.create(data);
        const obj = await createToken(newUser);
        sendMail(obj.token);
        return { success: 'success' };
    } catch(err) {
        return err
    }
}

async function updateUser(data) {
    console.log('updateUser:', data);
    try {
        const user = await User.findById(data.id);
        console.log('User:', user);
        user.firstName = data.firstName;
        user.lastName = data.lastName;
        return await user.save()
    } catch (e) {
        console.log('findUser error', e);
        return e;
    }
}

async function userVerified(id) {
    try {
        const user = await User.findById(id);
        user.isVerified = true;
        return await user.save()
    } catch (e) {
        console.log('userVerified error', e);
        return e;
    }
}

async function resetPassword(token, password) {
    try {
        const id = await getUserId(token);
        const user = await User.findById(id);
        user.password = password;
        await user.save();

        // Don not delete token when running functional tests
        if (!process.env.TEST) {
            await deleteToken(token);
        }

        return {
            confirmation: 'New password is saved'
        }
    } catch (e) {
        console.log('resetPassword error', e);
        return e;
    }
}

module.exports = {
    findUser,
    createNewUser,
    updateUser,
    userVerified,
    findEmail,
    resetPassword,
};
