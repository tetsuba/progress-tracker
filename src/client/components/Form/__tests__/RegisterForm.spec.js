import {
  delay,
  graphRenderer,
  updateTextInput,
} from '../../../../test/testHelper'
import RegisterForm from '../RegisterForm'
import { act } from 'react-dom/test-utils'
import { registerMockDataSuccess } from '../../../../test/mockApi/user/userMockMutation'

describe('<RegisterForm>', () => {
  describe('@Render', () => {
    it('should render', async () => {
      let wrapper
      await act(async () => {
        wrapper = graphRenderer(RegisterForm, [], {})
        await delay()
      })
      wrapper.update()
      expect(wrapper.find(RegisterForm)).toMatchSnapshot()
    })
  })

  describe('@Events', () => {
    describe('onSubmit', () => {
      it('should call "setPageState"', async () => {
        const props = {
          setPageState: jest.fn(),
        }

        const wrapper = graphRenderer(
          RegisterForm,
          [registerMockDataSuccess],
          props
        )

        act(() => {
          updateTextInput(wrapper, 'firstName', 'unit')
        })

        wrapper.update()

        act(() => {
          updateTextInput(wrapper, 'lastName', 'test')
        })

        wrapper.update()

        act(() => {
          updateTextInput(wrapper, 'email', 'unit@test.com')
        })

        wrapper.update()

        act(() => {
          updateTextInput(wrapper, 'newPassword', '1234')
        })

        wrapper.update()

        act(() => {
          updateTextInput(wrapper, 'confirmPassword', '1234')
        })

        wrapper.update()

        await act(async () => {
          await wrapper.find('#RegisterForm').get(0).props.onSubmit({
            preventDefault: jest.fn(),
          })
          await delay()
        })

        wrapper.update()

        expect(props.setPageState).toHaveBeenCalledTimes(1)
      })
    })
  })
})
