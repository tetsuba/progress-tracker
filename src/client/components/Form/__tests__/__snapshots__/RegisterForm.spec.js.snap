// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterForm> @Render should render 1`] = `
<Box
  max={500}
>
  <h3>
    Registration
  </h3>
  <Form
    className="mt-4 mb-5"
    id="RegisterForm"
    inline={false}
    onSubmit={[Function]}
  >
    <FormRow>
      <FormGroup
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "Col",
            "render": [Function],
          }
        }
        controlId="RegisterFirstName"
      >
        <FormControl
          name="firstName"
          onChange={[Function]}
          placeholder="First Name"
          required={true}
          value=""
        />
      </FormGroup>
    </FormRow>
    <FormRow>
      <FormGroup
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "Col",
            "render": [Function],
          }
        }
        controlId="RegisterLastName"
      >
        <FormControl
          name="lastName"
          onChange={[Function]}
          placeholder="Last Name"
          required={true}
          value=""
        />
      </FormGroup>
    </FormRow>
    <FormRow
      className="mb-3"
    >
      <FormGroup
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "Col",
            "render": [Function],
          }
        }
        controlId="RegisterEmail"
      >
        <FormControl
          isInvalid={false}
          name="email"
          onChange={[Function]}
          placeholder="Your email address"
          required={true}
          type="email"
          value=""
        />
        <Feedback
          type="invalid"
        />
      </FormGroup>
    </FormRow>
    <PasswordInputs
      inputs={
        Object {
          "confirmPassword": "",
          "email": "",
          "firstName": "",
          "lastName": "",
          "newPassword": "",
        }
      }
      register={true}
      setInputs={[Function]}
    />
    <Button
      active={false}
      className="float-right"
      disabled={false}
      id="RegisterFormSubmit"
      type="submit"
      variant="primary"
    >
      Register
    </Button>
  </Form>
</Box>
`;
